pkg/api/subscribeWorker.go

```diff
@@ -0,0 +1,196 @@
+package api
+
+import (
+	"context"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"time"
+
+	"github.com/xmtp/xmtpd/pkg/db"
+	"github.com/xmtp/xmtpd/pkg/db/queries"
+	"github.com/xmtp/xmtpd/pkg/proto/xmtpv4/message_api"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	subscriptionBufferSize    = 1024
+	maxSubscriptionsPerClient = 10000
+	SubscribeWorkerPollTime   = 100 * time.Millisecond
+	subscribeWorkerPollRows   = 10000
+	maxTopicLength            = 128
+)
+
+type listener = chan<- []*message_api.OriginatorEnvelope
+
+// A worker that listens for new envelopes in the DB and sends them to subscribers
+// Assumes that there are many listeners - non-blocking updates are sent on buffered channels
+// and may be dropped if full
+type subscribeWorker struct {
+	ctx context.Context
+	log *zap.Logger
+
+	dbSubscription <-chan []queries.GatewayEnvelope
+	// Assumption: listeners cannot be in multiple slices
+	globalListeners     []listener
+	originatorListeners map[uint32][]listener
+	topicListeners      map[string][]listener
```

Comment:
This is a little different from the previous implementation. Storing it in this format allows the subscribeWorker to only iterate over relevant listeners for each payload, rather than iterating over all listeners.

However, for doing listener cleanup, I think there will need to be additional metadata on listeners in the reverse direction (storing the topics/originators for each listener, so that it can be cleaned from the relevant maps)